modules/dnn/src/model.cpp

```diff
@@ -0,0 +1,229 @@
+#include <algorithm>
+#include <iostream>
+#include <utility>
+#include <iterator>
+
+#include <opencv2/imgproc.hpp>
+
+namespace cv {
+namespace dnn {
+
+struct Model::Impl {
+    Size   size;
+    Scalar mean;
+    float  scale = 1.0;
+    bool   swapRB = false;
+    bool   crop = false;
+    Mat    blob;
+    std::vector<String> outNames;
+
+    void predict(Net& net, const Mat& frame, std::vector<Mat>& outs) {
+        CV_Assert(!size.empty());
+
+        blobFromImage(frame, blob, 1.0, size, Scalar(), swapRB, crop, CV_8U);
+        net.setInput(blob, "", scale, mean);
+
+        // Faster-RCNN or R-FCN
+        if (net.getLayer(0)->outputNameToIndex("im_info") != -1) {
+            Mat imInfo = (Mat_<float>(1, 3) << size.height, size.width, 1.6f);
+            net.setInput(imInfo, "im_info");
+        }
+        net.forward(outs, outNames);
+    }
+};
+
+Model::Model(const String& model, const String& config)
+    : Net(readNet(model, config)), impl(new Impl) {
+        impl->outNames = getUnconnectedOutLayersNames();
+    };
+
+Model::Model(const Net& network) : Net(network), impl(new Impl) {
+    impl->outNames = getUnconnectedOutLayersNames();
+};
+
+Ptr<Model> Model::create(const String& model, const String& config) {
+    return makePtr<Model>(model, config);
+}
+
+Model& Model::setInputSize(const Size& size) {
+    impl->size = size;
+    return *this;
+}
+
+Model& Model::setInputSize(int width, int height) {
+    impl->size = Size(width, height);
+    return *this;
+}
+
+Model& Model::setInputMean(const Scalar& mean) {
+    impl->mean = mean;
+    return *this;
+}
+
+Model& Model::setInputScale(float scale) {
+    impl->scale = scale;
+    return *this;
+}
+
+Model& Model::setInputCrop(bool crop) {
+    impl->crop = crop;
+    return *this;
+}
+
+Model& Model::setInputSwapRB(bool swapRB) {
+    impl->swapRB = swapRB;
+    return *this;
+}
+
+void Model::setInputParams(const Size& size, const Scalar& mean,
+                           float scale, bool swapRB, bool crop) {
+    impl->size = size;
+    impl->mean = mean;
+    impl->scale = scale;
+    impl->crop = crop;
+    impl->swapRB = swapRB;
+}
+
+void Model::predict(InputArray frame, OutputArrayOfArrays outs) {
+    std::vector<Mat> outputs;
+    outs.getMatVector(outputs);
+    impl->predict(*this, frame.getMat(), outputs);
+}
+
+ClassificationModel::ClassificationModel(const String& model, const String& config)
+    : Model(model, config) {};
+
+ClassificationModel::ClassificationModel(const Net& network) : Model(network) {};
+
+Ptr<ClassificationModel> ClassificationModel::create(const String& model, const String& config) {
+    return makePtr<ClassificationModel>(model, config);
+}
+
+std::pair<int, float> ClassificationModel::classify(InputArray frame) {
+    std::vector<Mat> outs;
+    impl->predict(*this, frame.getMat(), outs);
+    CV_Assert(outs.size() == 1);
+
+    double conf;
+    cv::Point maxLoc;
+    minMaxLoc(outs[0], nullptr, &conf, nullptr, &maxLoc);
```

Comment:
I think that we need to reshape `outs[0]` because it might has more than two dimensions and `minMaxLoc` will fail:

```cpp
minMaxLoc(outs[0].reshape(1, 1), nullptr, &conf, nullptr, &maxLoc);
```